### ISSUE TITLE
bughuntoor - If user merges their `veNFT`, they'll lose part of their rewards

### ISSUE THREAD

**Issue by sherlock-admin2** [sherlock-audit/2024-06-velocimeter-judging#235]
bughuntoor

High

# If user merges their `veNFT`, they'll lose part of their rewards

## Summary
If user merges their `veNFT`, they'll lose part of their rewards

## Vulnerability Detail
When users claim rewards, they can at most claim up to the week before `last_token_time`.
```solidity
        for (uint i = 0; i < 50; i++) {
            if (week_cursor >= _last_token_time) break;
```
And given that `last_token_time` can at most be this week, this means that rewards in the `RewardsDistributor` are lagging at least a week at a time.

Then, if we look at the code of `merge` we'll see that the `from` token is actually burned.

```solidity
    function merge(uint _from, uint _to) external {
        require(attachments[_from] == 0 && !voted[_from], "attached");
        require(_from != _to);
        require(_isApprovedOrOwner(msg.sender, _from));
        require(_isApprovedOrOwner(msg.sender, _to));

        LockedBalance memory _locked0 = locked[_from];
        LockedBalance memory _locked1 = locked[_to];
        uint value0 = uint(int256(_locked0.amount));
        uint end = _locked0.end >= _locked1.end ? _locked0.end : _locked1.end;

        locked[_from] = LockedBalance(0, 0);
        _checkpoint(_from, _locked0, LockedBalance(0, 0));
        _burn(_from);
        _deposit_for(_to, value0, end, _locked1, DepositType.MERGE_TYPE);
    }
```

Since `claim` requires `msg.sender` to be approved or owner, because the token is burned, they won't be able to claim the rewards.
Any time a user merges their `veNFT`, they'll lose at least 1 week of rewards.

## Impact
Loss of funds

## Code Snippet
https://github.com/sherlock-audit/2024-06-velocimeter/blob/main/v4-contracts/contracts/VotingEscrow.sol#L1208

## Tool used

Manual Review

## Recommendation
Do not burn the token

**Comment by nevillehuang** [comment]
Note, please consider not making an escalation for the following issues and duplicates as I will make a self escalation to avoid a long drawn out escalation on multiple different issues. For any issues relating to duplicates of this issue, please leave comments here so we can aggregate comments and reconsider validity.

1. Most issues are likely invalid, user error, they can simply claim before withdrawing/merging, similar to this issue highlighted [here](https://github.com/sherlock-audit/2024-06-magicsea-judging/issues/283)
2. #235 and #236 makes the only valid point that the current week rewards are lost as rewards are lagging by one week, and is the only one that mentions the valid attack path, so I believe it is the only issue that is valid.

- #170, #367, #606, #682 - Some has good PoCs, but unfortunately, does not identify the attack path mentioned in point 1 above
- #236 - Valid, but would consider dupe of #235 because it has the same root cause per [sherlock duplication guidelines](https://github.com/sherlock-protocol/sherlock-v2-docs/tree/29a019b8690eb01df3587868c4d7eac1e3459bda/audits/judging/judging#ix-duplication-rules) and mentioned the lagging rewards

**Comment by nevillehuang** [comment]
Escalate as per comments above and as discussed [here](https://discord.com/channels/812037309376495636/1257350045976760404/1272478338610499626)

**Comment by sherlock-admin3** [comment]
> Escalate as per comments above and as discussed [here](https://discord.com/channels/812037309376495636/1257350045976760404/1272478338610499626)

You've created a valid escalation!

To remove the escalation from consideration: Delete your comment.

You may delete or edit your escalation comment anytime before the 48-hour escalation window closes. After that, the escalation becomes final.

**Comment by sonny2k** [comment]
Pointing out 1 week of lagging rewards seems like it's more of Impact elaboration rather than Attack Path elaboration itself, just more detailed than other issues as other issues simply state unclaimed rewards are lost. I think this impact elaboration is obvious, as long as the rewards are not claimed before calling merge/withdraw, the lost unclaimed rewards will be in the range of [1 week - 51 weeks]. Making this issue not so different from its dups, just more specific on the impact ofc. So if this issue is valid then all of its dups should be also valid IMO. BTW thanks for your hard work @nevillehuang! Your work judging this contest is truly sonorous.

**Comment by 0xklapouchy** [comment]
@nevillehuang

Correct me if I'm wrong, but due to the `rewards lagging by one week`, when using the merge (the current week’s rewards from one tokenId are transferred to another tokenId), the rewards are actually moved to the new tokenId. There is no reward loss for the current week. (rewards can even increase if `to` tokenId has a longer endTime).

Rewards are calculated based on the balanceOf at each `week_cursor`, with no rewards calculations occurring in between. The lock in the VotingEscrow is also based on weeks, so you can’t withdraw() before the entire week has passed.

Therefore, the assumption that `current week rewards are lost` is invalid.

Based on the issue discussed at https://github.com/sherlock-audit/2024-06-magicsea-judging/issues/283, you should either invalidate all issues and classify them as `Low`, or determine that all of them are valid.

**Comment by spacegliderrrr** [comment]
Statement above is incorrect. When merging, current week’s rewards are not transferred to new token. 

During week N, user can claim rewards up to week `N - 1`, based on their balance at the beginning of week `N - 1`. Rewards for week N will be lost, as the token will be burned, and the new tokenId will receive the amount after the week has started (so the amount will be accounted for from the next week onward)

**Comment by 0xklapouchy** [comment]
I think proper PoC is required, to my understanding this code:

```solidity
File: RewardsDistributorV2.sol
198:             if (week_cursor >= user_point.ts && user_epoch <= max_user_epoch) {
199:                 user_epoch += 1;
200:                 old_user_point = user_point;
201:                 if (user_epoch > max_user_epoch) {
202:                     user_point = IVotingEscrow.Point(0,0,0,0);
203:                 } else {
204:                     user_point = IVotingEscrow(ve).user_point_history(_tokenId, user_epoch);
205:                 }
206:             } else {
```

Will look for the `last user point history` for the week `N-1`, therefore not the balance at the `beginning` but at the `end` of the week is used. (as you are looking util the week_cursor is less then user_point.ts, this indicates that else is called on the last known user_point from that week_cursor)

Before the merge user can claim his `N-1` rewards. If not claimed it is user error.

And as I stated before during the merge the new tokenId balance is updated for the `N` week, the `last user point history` for `N` week_cursor for that week will contain the amount from merged tokenId therefore the rewards will be there.

**Comment by nevillehuang** [comment]
1. I disagree that the precondition for attack path is not important, because without it, I would have invalidate all issues as user error since without me actually going and find the actual vulnerability path myself, there was no way I would have known that rewards of a lagging week will be lost
2. Regarding @0xklapouchy claims, I would need more code/example logic to determine if it is correct. From my understanding since the `_burn` was first invoked [here](https://github.com/sherlock-audit/2024-06-velocimeter/blob/main/v4-contracts/contracts/VotingEscrow.sol#L1208), the rewards will be lost .

**Comment by cvetanovv** [comment]
I agree with @sonny2k comment. All duplicates have stated a root cause, and it is that rewards are lost when merging and withdrawing. 

They have also pointed out the impact: reward will be lost.

Because the escalation is from Lead Judge for discussion purposes and this issue will remain valid, I plan to reject the escalation but duplicate #170, #367, #606, and #682 with this issue.

**Comment by WangSecurity** [comment]
Result:
High
Has duplicates

**Comment by sherlock-admin4** [comment]
Escalations have been resolved successfully!

Escalation status:
- [nevillehuang](https://github.com/sherlock-audit/2024-06-velocimeter-judging/issues/235/#issuecomment-2283436094): rejected

**Comment by sherlock-admin2** [comment]
The protocol team fixed this issue in the following PRs/commits:
https://github.com/Velocimeter/v4-contracts/pull/24

**Comment by spacegliderrrr** [comment]
Fix looks good. Upon burning a token, the last owner is saved in mapping, which later the `RewardsDistributor` and `Bribe` check.

**Comment by sherlock-admin2** [comment]
The Lead Senior Watson signed off on the fix.
